subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientProvidedBuildActionRunner.java

```diff
@@ -44,22 +43,14 @@ public void run(BuildAction action, final BuildController buildController) {
         }
 
         final GradleInternal gradle = buildController.getGradle();
+        gradle.getStartParameter().setConfigureOnDemand(false);
```

Comment:
A build action allows you to request several models in one go. You can even have conditional logic, looking at one model and then deciding what other models to get. That's much more efficient than the `getModel()` method on `ProjectConnection` where each call will result in configuring the project again.

Apart from that you can think of them as the same thing: You are trying to get models out of a project and for that to work, all projects need to be fully configured. Since they need to be fully configured anyway, there is no point in using "configure on demand". Using it leads to bugs like `buildFinished` hooks never being called.